src/transform/src/attribute/column_names.rs

```diff
@@ -0,0 +1,265 @@
+                input_results.next().unwrap().clone()
+            }
+            Join {
+                inputs: _,
+                equivalences: _,
+                implementation: _,
+            } => {
+                let mut column_names = vec![];
+                for input_column_names in input_results {
+                    column_names.extend(input_column_names.iter().cloned());
+                }
+                column_names
+            }
+            Reduce {
+                input: _,
+                group_key,
+                aggregates,
+                monotonic: _,
+                expected_group_size: _,
+            } => {
+                // We clone and extend the input vector and then remove the part
+                // associated with the input at the end.
+                let mut column_names = input_results.next().unwrap().clone();
+                let input_arity = column_names.len();
+
+                // Infer the group key part.
+                ColumnNames::extend_with_scalars(&mut column_names, group_key);
+                // Infer the aggregates part.
+                let aggs_start = group_key.len();
+                let aggs_end = group_key.len() + aggregates.len();
+                column_names.extend(ColumnNames::anonymous(aggs_start..aggs_end));
+                // Remove the prefix associated with the input
+                column_names.drain(0..input_arity);
+
+                column_names
+            }
+            TopK {
+                input: _,
+                group_key: _,
+                order_key: _,
+                limit: _,
+                offset: _,
+                monotonic: _,
+                expected_group_size: _,
+            } => {
+                // Return the column names of the `input`.
+                input_results.next().unwrap().clone()
+            }
+            Negate { input: _ } => {
+                // Return the column names of the `input`.
+                input_results.next().unwrap().clone()
+            }
+            Threshold { input: _ } => {
+                // Return the column names of the `input`.
+                input_results.next().unwrap().clone()
+            }
+            Union { base: _, inputs: _ } => {
+                // Use the first non-empty column across all inputs.
+                let mut column_names = vec![];
+
+                let base_results = input_results.next().unwrap();
+                let inputs_results = input_results.collect::<Vec<_>>();
+
+                for (i, base_column_name) in base_results.iter().enumerate() {
+                    let mut column_name = base_column_name.clone();
+                    for input_results in inputs_results.iter() {
+                        if !input_results[i].is_empty() {
+                            column_name = input_results[i].clone();
+                            break;
+                        }
+                    }
+                    column_names.push(column_name);
```

Comment:
Not sure if I understand the MIR `Union` semantics 100% vs. the logic here: If we have a `base`, then wouldn't the column names there always be ignored if there are other inputs with column names present? Is that intended?

For example, consider a test case in which a `Union` would have a `base` and two `inputs` s.t. the `base` and the last of `inputs` have all column names given. Wouldn't the fragment above take the column names from the last of `inputs` and not from `base`?

Reply:
Yes, currently the last input wins. I can change this to "first input wins" semantics, but most of the time I am expecting identical input positions to also resolve to identical column names (at least in SQL they are required to align). 

I don't think there is a silver bullet solution here. Consider the following queries:

```sql
CREATE TABLE R(x int, y int);
CREATE TABLE S(y int, x int);

-- Q1
EXPLAIN WITH(column_names)
SELECT x as y, y as x FROM R 
UNION ALL
SELECT y, x FROM S;

-- Q2
EXPLAIN WITH(column_names)
SELECT x, y FROM R 
UNION ALL
SELECT y as x, x as y FROM S;
```

Currently, in both cases we will infer `(y, x)`, which is bad for Q2 where the schema is actually `(x, y)`. If I change the behavior it in this degenerate case things will get bad for Q1, but I think it's probably the more intuitive solution.

Reply:
What I'm finding confusing is that actually it is not that the last input wins, but the first with names that is different than `base`. Let me augment your example a bit:
```sql
CREATE TABLE R(x int, y int);
CREATE TABLE S(y int, x int);
CREATE TABLE T(a int, b int);
CREATE TABLE U(c int, d int);

-- Q1
EXPLAIN WITH(column_names)
SELECT x as y, y as x FROM R 
UNION ALL
SELECT y, x FROM S;

-- Q2
EXPLAIN WITH(column_names)
SELECT x, y FROM R 
UNION ALL
SELECT y as x, x as y FROM S;

-- Q3
EXPLAIN WITH(column_names)
SELECT x as y, y as x FROM R 
UNION ALL
SELECT y, x FROM S
UNION ALL
SELECT a, b FROM T;

-- Q4
EXPLAIN WITH(column_names)
SELECT x, y FROM R 
UNION ALL
SELECT a, b FROM T
UNION ALL
SELECT y as x, x as y FROM S;

-- Q5
EXPLAIN WITH(column_names)
SELECT x, y FROM R 
UNION ALL
SELECT a + 1, b + 1 FROM T
UNION ALL
SELECT c, d FROM U
UNION ALL
SELECT y as x, x as y FROM S;
```
What we get as a result for the `EXPLAIN`s is:
```
                           Optimized Plan                           
--------------------------------------------------------------------
 Explained Query:                                                  +
   Union // { column_names: "(y, x)" }                             +
     ReadStorage materialize.public.r // { column_names: "(x, y)" }+
     ReadStorage materialize.public.s // { column_names: "(y, x)" }+
 
(1 row)

                           Optimized Plan                           
--------------------------------------------------------------------
 Explained Query:                                                  +
   Union // { column_names: "(y, x)" }                             +
     ReadStorage materialize.public.r // { column_names: "(x, y)" }+
     ReadStorage materialize.public.s // { column_names: "(y, x)" }+
 
(1 row)

                           Optimized Plan                           
--------------------------------------------------------------------
 Explained Query:                                                  +
   Union // { column_names: "(y, x)" }                             +
     ReadStorage materialize.public.r // { column_names: "(x, y)" }+
     ReadStorage materialize.public.s // { column_names: "(y, x)" }+
     ReadStorage materialize.public.t // { column_names: "(a, b)" }+
 
(1 row)

                           Optimized Plan                           
--------------------------------------------------------------------
 Explained Query:                                                  +
   Union // { column_names: "(a, b)" }                             +
     ReadStorage materialize.public.r // { column_names: "(x, y)" }+
     ReadStorage materialize.public.t // { column_names: "(a, b)" }+
     ReadStorage materialize.public.s // { column_names: "(y, x)" }+
 
(1 row)

                             Optimized Plan                             
------------------------------------------------------------------------
 Explained Query:                                                      +
   Union // { column_names: "(c, d)" }                                 +
     ReadStorage materialize.public.r // { column_names: "(x, y)" }    +
     Project (#2, #3) // { column_names: "(#0, #1)" }                  +
       Map ((#0 + 1), (#1 + 1)) // { column_names: "(a, b, #2, #3)" }  +
         ReadStorage materialize.public.t // { column_names: "(a, b)" }+
     ReadStorage materialize.public.u // { column_names: "(c, d)" }    +
     ReadStorage materialize.public.s // { column_names: "(y, x)" }    +
 
(1 row)
```

Reply:
I change that code to reflect what the comment says:

```rust
// Use the first non-empty column across all inputs.
let mut column_names = vec![];

let base_results = input_results.next().unwrap();
let inputs_results = input_results.collect::<Vec<_>>();

for (i, mut column_name) in base_results.iter().cloned().enumerate() {
    for input_results in inputs_results.iter() {
        if column_name.is_empty() && !input_results[i].is_empty() {
            column_name = input_results[i].clone();
            break;
        }
    }
    column_names.push(column_name);
}
```

Reply:
Perfect, thanks!